File: src/main/java/leetcode/hashtable/LoggerRateLimiter.java
package leetcode.hashtable;

import java.util.HashMap;
import java.util.Map;

public class LoggerRateLimiter {
  /**
   * Initialize your data structure here.
   */
  Map<String, Integer> timeTracker = new HashMap<>();

  public LoggerRateLimiter() {

  }

  /**
   * Returns true if the message should be printed in the given timestamp, otherwise returns false.
   * If this method returns false, the message will not be printed.
   * The timestamp is in seconds granularity.
   */
  public boolean shouldPrintMessage(int timestamp, String message) {
    if (!timeTracker.containsKey(message)) {
      timeTracker.put(message, timestamp);
      return true;
    } else {
      int lastLoggedTime = timeTracker.get(message);
      if (timestamp - lastLoggedTime >= 10) {
        timeTracker.put(message, timestamp);
        return true;
      }
      return false;
    }
  }
}

File: src/main/go/hashtable/logger_rate_limiter.go
package hashtable

type Logger struct {
	tracker map[string]int
}

/** Initialize your data structure here. */
func Constructor() Logger {
	return Logger{tracker: map[string]int{}}
}

/** Returns true if the message should be printed in the given timestamp, otherwise returns false.
  If this method returns false, the message will not be printed.
  The timestamp is in seconds granularity. */
func (this *Logger) ShouldPrintMessage(timestamp int, message string) bool {
	if _, ok := this.tracker[message]; !ok {
		this.tracker[message] = timestamp
		return true
	} else {
		lastLogged, _ := this.tracker[message]
		if timestamp-lastLogged >= 10 {
			this.tracker[message] = timestamp
			return true
		}
		return false
	}
}

/**
 * Your Logger object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.ShouldPrintMessage(timestamp,message);
 */
